### PR TITLE
New design policy proposal

### PR DESCRIPTION
**Pull request by hlandau**

### CODE REVIEW

File: policies/design-process.md [hunk 1]
```diff
@@ -1,160 +1,164 @@
...
+Requirement for Design Documents
+--------------------------------
+
+A design document is required for any proposed enhancement which adds new APIs
+or evolves or modifies existing APIs.
+
+For any other kind of proposed enhancement, a design document should be created
+if it incorporates design decisions or aspects significant enough to warrant
+one. For example, if an enhancement adds a new internal module with clearly
+delineated boundaries with an documented internal API which can be consumed by
```
Comment (paulidale) [inline policies/design-process.md:50]: Just _a_ rather than _an_.

**Review by paulidale — COMMENTED** [review]
Looks good.

File: policies/design-process.md [hunk 2]
```diff
@@ -1,160 +1,164 @@
...
+   A presentation to the OTC may be made but is not required.
+
+When a design is produced, it should be submitted to the OpenSSL repository as a
+PR. It should be determined which level of scrutiny is appropriate according to
+this policy and this should be noted in the PR. Any applicable actions (such as
+notifying the OTC via the list) should be carried out once the design is ready
+for review.
+
+Any OTC member may object to the processing of a design at a given level of
+scrutiny. In this case, the next highest level of scrutiny is used. An OTC
```
Comment (levitte) [inline policies/design-process.md:92]: s/next highest level/next higher level/

**Comment by levitte** [comment]
I like what I see.  I'll have another read tomorrow.

File: policies/design-process.md (hunk 2 above)
Comment (paulidale) [reply policies/design-process.md:92]: I think this should be rewritten to say that any OTC member can object and request processing at any higher level.

Being able to go to the next higher or the top level is effectively the same thing but more wordy and less streamlined IMO.  It's also less future proof if we ever add more review levels (which I really hope never happens).

File: policies/design-process.md [hunk 3]
```diff
@@ -1,160 +1,164 @@
...
+It is recognised that designs will often necessarily change once implementation
+begins. In the majority of cases, the understanding of the problem domain will
+evolve and improve after implementation begins and this will indicate further
+changes to the design. This is an iterative process of progressive refinement.
+It is an explicit objective of this policy to support and encourage this
+agile-style process, as opposed to a waterfall-style process in which designs
+must be approved and finalised prior to implementation.
+
+Moreover, it may be hard to obtain explicit approval for a design from the OTC
+where there design falls out of the domains of the OTC members' expertise. This
```
Comment (t8m) [inline policies/design-process.md:33]: I would just drop this sentence as there are also other reasons why graded system of reviews is good idea.

File: policies/design-process.md [hunk 4]
```diff
@@ -1,160 +1,164 @@
...
+A proposer may choose to use a higher level of scrutiny than is required.
+
+Selecting a Level
+-----------------
+
+To determine the level of scrutiny which must be applied to a design by default,
+follow the following process:
+
+  - Any design which proposes to create new public APIs, or evolve or modify
+    existing public APIs, must use at least the Present level of scrutiny.
```
Comment (t8m) [inline policies/design-process.md:104]: Again I'd add `non-trivially evolve or modify`

File: policies/design-process.md [hunk 5]
```diff
@@ -1,160 +1,164 @@
...
+
+  - Any other design may use the Notify level of scrutiny.
+
+Checklists
+----------
+
+### Checklist for the Notify Level
+
+  - Design document published as a PR on GitHub
+  - OTC notified and invited to comment or object via an email to OTC list
```
Comment (t8m) [inline policies/design-process.md:114]: openssl-project list

File: policies/design-process.md [hunk 6]
```diff
@@ -1,160 +1,164 @@
...
+  - At least one week has passed from OTC notification
+
+### Checklist for the Present Level
+
+  - Design document published as a PR on GitHub
+  - OTC notified and invited to comment or object via an email to OTC list
```
Comment (t8m) [inline policies/design-process.md:120]: as above

**Review by t8m — CHANGES_REQUESTED** [review]
Overall this should work much better than the previous design policy. Nice work!

**Comment by hlandau** [comment]
Minor updates.

File: policies/design-process.md [hunk 7]
```diff
@@ -1,160 +1,167 @@
...
+   A minimum waiting time of one week applies to ensure OTC members have the
+   opportunity to review the design.
+
+   The notification does not need to be made by the author of the design
+   document.
+
+ - At the **Present** level, the OTC is notified of a new design
+   in the same way that it is at the Notify level. The same minimum waiting time
+   applies. The design is also introduced and explained in a presentation given
+   by its author to the OTC in a meeting of the OTC. The OTC has opportunities
```
Comment (mattcaswell) [inline policies/design-process.md:78]: I don't think it necessarily has to be the author - although that would be normal. For example, in the case of an external contribution someone might present the design on the author's behalf (e.g. if the OTC meeting is in the middle of the night for the contributor)

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: policies/design-process.md [hunk 8]
```diff
@@ -1,160 +1,167 @@
...
+ - Notify
+ - Present
+ - Approve
+
+These levels work as follows:
+
+ - At the **Notify** level, the OTC is notified of a new design
+   when it is available for review, via an email to the openssl-project mailing list.
+   OTC members and committers can review and comment on the design.
+   A minimum waiting time of one week applies to ensure OTC members have the
```
[the remark this answers is not in the document]
Comment (hlandau) [reply policies/design-process.md:69]: I was thinking from time of notification (see checklist).

File: policies/design-process.md [hunk 9]
```diff
@@ -1,160 +1,167 @@
...
+below in ascending order of severity:
+
+ - Notify
+ - Present
+ - Approve
+
+These levels work as follows:
+
+ - At the **Notify** level, the OTC is notified of a new design
+   when it is available for review, via an email to the openssl-project mailing list.
```
[the remark this answers is not in the document]
Comment (hlandau) [reply policies/design-process.md:67]: Either could work, though I suspect GH notifications have a lower signal to noise ratio. Seeking more opinions before changing this.

**Comment by hlandau** [comment]
Minor updates.

**Comment by hlandau** [comment]
Updated to add discussion of what we want in design documents.

This update was inspired by the way the Vulkan community produces designs, in particular:
- The use of Q&A sections to provide 'Socratic' exposition and historical documentation of design decisions, examples of which can be found here: https://github.com/KhronosGroup/Vulkan-Docs/search?q=issues+resolved
- Their template for design documents: https://github.com/KhronosGroup/Vulkan-Docs/blob/main/proposals/template.adoc

**Comment by hlandau** [comment]
@t-j-h Thoughts welcomed.

File: policies/design-process.md [hunk 10]
```diff
@@ -1,160 +1,235 @@
...
-
-It should be possible for community members to see the progress of designs
-through the process and to pass comment on the proposed design prior to its
-eventual agreement by OTC. As a minimum:
-
-* The draft design document containing the candidate solutions should be
-available as a pull request for public comment for at least one week prior to
-the selection of the solution. In addition to that, community members may be
-offered an invite to join the solution selection meeting.
+Design Process Policy
```
Comment (levitte) [inline policies/design-process.md:1]: Could we remove "Policy" from the title, please?  Or do we not know that this is a policy?

The result on our web site is a blur of "Polify for \<blah\>" and "\<blah\> Policy", which makes it harder to quickly find what you're looking for, visually speaking.

https://www.openssl.org/policies/technical/

**Comment by t8m** [comment]
Please review so this can be voted.

File: policies/design-process.md (hunk 10 above)
Comment (paulidale) [reply policies/design-process.md:1]: Currently we have no standard for naming, we need one but an ad-hoc decision here isn't the right path IMO.  A consistent naming convention that is applied everywhere is the right way.

There is an old ticket to rectify this but it's not been considered high enough priority to be acted on.

Comment (levitte) [reply policies/design-process.md:1]: I think it should be lifted on the priority ladder, then

Comment (levitte) [reply policies/design-process.md:1]: ... also, the choice of title is pretty ad-hoc already, so you know, trading one ad-hoc for another...?

Comment (paulidale) [reply policies/design-process.md:1]: ...but your ad-hoc call is clearly as incorrect as @hlandau's.  It should be called: _Machiavellian Frobnications and Design_ :laughing:

Comment (t8m) [reply policies/design-process.md:1]: Or _How to Choose the Proper Color for a Bikeshed_

**Comment by t8m** [comment]
**OTC: Any substantial comments should be submitted by Apr 4 and the voting on this will start on Apr 11 after these comments are resolved.**

**Comment by levitte** [comment]
I still like what I see.

The only thing I might have an slight issue with is the phrase "Implementation should generally begin immediately.", if *should* is interpreted too strongly...  but that can be debated later, if the issue actually arises.

**Comment by paulidale** [comment]
We could change _should generally_ to _can_ which weakens things.  I agree that this tends to indicate that implementation starts now which isn't ideal.

File: policies/design-process.md [hunk 11]
```diff
@@ -1,160 +1,235 @@
...
+ - Present
+ - Approve
 
-* The refined documented design should be available for public comment as a pull
-request for at least one week prior to eventual agreement of the solution
+These levels work as follows:
+
+ - At the **Notify** level, the OTC is notified of a new design
+   when it is available for review, via an email to the openssl-project mailing list.
+   OTC members and committers can review and comment on the design.
```
Comment (slontis) [inline policies/design-process.md:135]: Why is restricted to OTC and committers?

Comment (hlandau) [reply policies/design-process.md:135]: This wording isn't intended to imply any such limitation... I'll tweak it.

**Comment by hlandau** [comment]
Slight updates.

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Opening the vote on the design policy proposal.

```
topic: Accept the design policy proposal (PR #58)
comment: https://github.com/openssl/technical-policies/pull/58
Proposed by Hugo.
Public: yes
opened: 2023-04-18
closed: 
accepted:    (for: ?, against: ?, abstained: ?, not voted: ?)

  Dmitry     [  ]
  Matt       [  ]
  Pauli      [  ]
  Tim        [  ]
  Hugo       [+1]
  Richard    [  ]
  Shane      [  ]
  Tomas      [  ]
  Kurt       [  ]
  Matthias   [  ]
  Nicola     [  ]
```

**Comment by t8m** [comment]
Vote: [+1]

**Comment by paulidale** [comment]
Vote: [0]

**Comment by romen** [comment]
Vote: [0]

**Comment by t-j-h** [comment]
Vote [+1]

**Comment by slontis** [comment]
Vote [+1]

**Comment by beldmit** [comment]
Vote: [0]

**Comment by hlandau** [comment]
Vote passes.

**Comment by levitte** [comment]
Vote: [+1]

**Comment by hlandau** [comment]
Merged.

**Comment by mattcaswell** [comment]
Vote: [+1]

**Comment by hlandau** [comment]
@mattcaswell Added.

**Pull request closed without merging**